src/scheduler.cpp

```diff
@@ -35,7 +35,7 @@ void Scheduler::threadMain()
 		}
 
 		// the mutex is locked again now...
-		if (ret == std::cv_status::timeout) {
+		if ((ret == std::cv_status::timeout) && !eventList.empty()) {
 			// ok we had a timeout, so there has to be an event we have to execute...
```

Comment:
What about this comment? Is it wrong, then?

Reply:
I made a mistake, i quick copied the fix from own src and its why there this extra "if (ret == std::cv_status::timeout) {" that will prevent the source to be compiled(its why the pr has two commits, the second one is to fix the stupid mistake in the first one). About the comment "the mutex is locked again now..." its not wrong. The problem lies on "eventSignal.wait_until". in the documentation http://en.cppreference.com/w/cpp/thread/condition_variable/wait_until its says "The thread will be unblocked when notify_all() or notify_one() is executed, or when the absolute time point timeout_time is reached. It may also be unblocked spuriously. When unblocked, regardless of the reason, lock is reacquired and wait_until exits." So while the eventSignal.wait_until is waiting the cycles the mutex is unlocked and then its possible to Scheduler::shutdown() clean both lists(eventIds and eventList).

Reply:
In tfs 1.1 there is an extra check that does not exists in tfs 1.2 code: "if (ret == std::cv_status::timeout && threadState != THREAD_STATE_TERMINATED) {" also in tfs 0.4 code: "if(!action && Scheduler::m_threadState != Scheduler::STATE_TERMINATED)" Maybe the old code was made to prevent this issue and in some rework it has been removed? @djarek

Reply:
I'm too lazy to check the history, but this might've been broken when code related to thread management was abstracted away.